File: day01/sol.py
from timing_util import Timing


def get_data(content: str) -> list[str]:
    return content.split("\n")


def calibration_number(line: str) -> int:
    def parse_number(line: str) -> int:
        for char in line:
            if char.isdigit():
                return int(char)

    first = parse_number(line)
    last = parse_number(reversed(line))
    return first * 10 + last


def part1(data: list[str]) -> int:
    return sum(map(calibration_number, data))


def calibration_word(line: str) -> int:
    def parse_number_left(line: str) -> int:
        chunk = ""
        for char in line:
            # if we encounter a number: return it immediately
            if char.isdigit():
                return int(char)

            chunk = chunk[-4:] + char

            if len(chunk) >= 3:
                if chunk.endswith("one"):
                    return 1
                if chunk.endswith("two"):
                    return 2
                if chunk.endswith("six"):
                    return 6
            if len(chunk) >= 4:
                if chunk.endswith("four"):
                    return 4
                if chunk.endswith("five"):
                    return 5
                if chunk.endswith("nine"):
                    return 9
            if chunk == "three":
                return 3
            if chunk == "seven":
                return 7
            if chunk == "eight":
                return 8

    def parse_number_right(line: str) -> int:
        chunk = ""
        for char in line:
            # if we encounter a number: return it immediately
            if char.isdigit():
                return int(char)

            chunk = char + chunk[:4]

            if len(chunk) >= 3:
                if chunk.startswith("one"):
                    return 1
                if chunk.startswith("two"):
                    return 2
                if chunk.startswith("six"):
                    return 6
            if len(chunk) >= 4:
                if chunk.startswith("four"):
                    return 4
                if chunk.startswith("five"):
                    return 5
                if chunk.startswith("nine"):
                    return 9
            if chunk == "three":
                return 3
            if chunk == "seven":
                return 7
            if chunk == "eight":
                return 8

    first = parse_number_left(line)
    second = parse_number_right(reversed(line))
    return first * 10 + second


def part2(data: list[str]) -> int:
    return sum(map(calibration_word, data))


if __name__ == "__main__":
    with Timing("Elapsed time: "):
        with open("inputs/day01.txt") as in_file:
            data = get_data(in_file.read().strip())

        print(f"part1: {part1(data)}")
        print(f"part2: {part2(data)}")

File: day01/sol.go
package main

import (
	"errors"
	"fmt"
	"os"
	"strings"
	"time"
)

func ReadFile(filename string) string {
	bytes, err := os.ReadFile(filename)
	if err != nil {
		panic(err)
	}
	return strings.TrimSpace(string(bytes))
}

func ParseInput(data string) []string {
	return strings.Split(data, "\n")
}

func Reverse(s string) string {
	n := len(s)
	chunk := make([]rune, n)
	for _, rune := range s {
		n--
		chunk[n] = rune
	}
	return string(chunk[n:])
}

func IsDigit(r rune) bool {
	return '0' <= r && r <= '9'
}

func ParseInt(r rune) int {
	return int(r - '0')
}

func CalibrationNumber(line string) int {
	for _, r := range line {
		if IsDigit(r) {
			return ParseInt(r)
		}
	}
	return -1
}

func Part1(lines []string) int {
	sum := 0
	for _, line := range lines {
		first := CalibrationNumber(line)
		second := CalibrationNumber(Reverse(line))
		sum += first*10 + second
	}
	return sum
}

func ParseNumberLeft(line string) int {
	chunk := make([]rune, 0)
	for _, r := range line {
		if IsDigit(r) {
			return ParseInt(r)
		}

		if len(chunk) >= 5 {
			chunk = append(chunk[1:], r)
		} else {
			chunk = append(chunk, r)
		}

		str := string(chunk)
		if len(str) >= 3 {
			if strings.HasSuffix(str, "one") {
				return 1
			}
			if strings.HasSuffix(str, "two") {
				return 2
			}
			if strings.HasSuffix(str, "six") {
				return 6
			}
		}
		if len(str) >= 4 {
			if strings.HasSuffix(str, "four") {
				return 4
			}
			if strings.HasSuffix(str, "five") {
				return 5
			}
			if strings.HasSuffix(str, "nine") {
				return 9
			}
		}
		if str == "three" {
			return 3
		}
		if str == "seven" {
			return 7
		}
		if str == "eight" {
			return 8
		}
	}
	panic(errors.New("illegal state reached"))
}

func ParseNumberRight(line string) int {
	chunk := make([]rune, 0)
	for _, r := range line {
		if IsDigit(r) {
			return ParseInt(r)
		}

		if len(chunk) >= 5 {
			chunk = append([]rune{r}, chunk[:4]...)
		} else {
			chunk = append([]rune{r}, chunk...)
		}

		str := string(chunk)
		if len(str) >= 3 {
			if strings.HasPrefix(str, "one") {
				return 1
			}
			if strings.HasPrefix(str, "two") {
				return 2
			}
			if strings.HasPrefix(str, "six") {
				return 6
			}
		}
		if len(str) >= 4 {
			if strings.HasPrefix(str, "four") {
				return 4
			}
			if strings.HasPrefix(str, "five") {
				return 5
			}
			if strings.HasPrefix(str, "nine") {
				return 9
			}
		}
		if str == "three" {
			return 3
		}
		if str == "seven" {
			return 7
		}
		if str == "eight" {
			return 8
		}
	}
	panic(errors.New("illegal state reached"))
}

func Part2(lines []string) int {
	sum := 0
	for _, line := range lines {
		first := ParseNumberLeft(line)
		second := ParseNumberRight(Reverse(line))
		sum += first*10 + second
	}
	return sum
}

func PrintTime(start time.Time, msg string) {
	dur := float64(time.Since(start).Nanoseconds())
	switch {
	case dur < 1e3:
		fmt.Printf("%v %.3fns\n", msg, dur)
	case dur < 1e6:
		fmt.Printf("%v %.3fµs\n", msg, dur/1e3)
	case dur < 1e9:
		fmt.Printf("%v %.3fms\n", msg, dur/1e6)
	default:
		fmt.Printf("%v %.3fs\n", msg, dur/1e9)
	}
}

func main() {
	start := time.Now()
	defer PrintTime(start, "Elapsed time:")

	text := ReadFile("inputs/day01.txt")
	data := ParseInput(text)

	fmt.Printf("Part1: %d\n", Part1(data))
	fmt.Printf("Part2: %d\n", Part2(data))
}
